src/EditorFeatures/CSharpTest/UsePatternMatching/CSharpIsAndCastCheckCodeRefactoringProviderTests.cs

```diff
@@ -0,0 +1,502 @@
+}");
+        }
+
+        [Fact, Trait(Traits.Feature, Traits.Features.CodeActionsInlineTypeCheck)]
+        public async Task TestLambdaBody()
+        {
+            await TestInRegularAndScript1Async(
+@"
+using System;
+
+class TestFile
+{
+    int i;
+
+    void Foo(Func<bool> f) { }
+
+    bool M(object obj)
+        => Foo(() => [||]obj is TestFile && ((TestFile)obj).i > 0, () => obj is TestFile && ((TestFile)obj).i > 0);
+}",
+@"
+using System;
+
+class TestFile
+{
+    int i;
+
+    void Foo(Func<bool> f) { }
+
+    bool M(object obj)
+        => Foo(() => obj is TestFile {|Rename:file|} && file.i > 0, () => obj is TestFile && ((TestFile)obj).i > 0);
+}");
+        }
+
+        [Fact, Trait(Traits.Feature, Traits.Features.CodeActionsInlineTypeCheck)]
+        public async Task TestDefiniteAssignment1()
+        {
+            await TestInRegularAndScript1Async(
+@"class TestFile
+{
+    int i;
+    bool M(object obj)
+    {
+        if ([||]obj is TestFile)
+        {
+            M(((TestFile)obj).i);
+            M(((TestFile)obj).i);
+        }
+        else
+        {
+            M(((TestFile)obj).i);
+            M(((TestFile)obj).i);
+        }
+    }
+}",
+@"class TestFile
+{
+    int i;
+    bool M(object obj)
+    {
+        if (obj is TestFile {|Rename:file|})
+        {
+            M(file.i);
+            M(file.i);
+        }
+        else
+        {
+            M(((TestFile)obj).i);
+            M(((TestFile)obj).i);
+        }
+    }
+}");
+        }
+
+        [Fact, Trait(Traits.Feature, Traits.Features.CodeActionsInlineTypeCheck)]
+        public async Task TestDefiniteAssignment2()
+        {
+            await TestInRegularAndScript1Async(
+@"class TestFile
+{
+    int i;
+    bool M(object obj)
+    {
+        if (!([||]obj is TestFile))
+        {
+            M(((TestFile)obj).i);
+            M(((TestFile)obj).i);
+        }
+        else
+        {
+            M(((TestFile)obj).i);
+            M(((TestFile)obj).i);
+        }
+    }
+}",
+@"class TestFile
+{
+    int i;
+    bool M(object obj)
+    {
+        if (!(obj is TestFile {|Rename:file|}))
+        {
+            M(((TestFile)obj).i);
+            M(((TestFile)obj).i);
+        }
+        else
+        {
+            M(file.i);
+            M(file.i);
+        }
+    }
+}");
+        }
+
+        [Fact, Trait(Traits.Feature, Traits.Features.CodeActionsInlineTypeCheck)]
+        public async Task TestNotOnAnalyzerMatch()
+        {
+            await TestMissingAsync(
+@"class TestFile
+{
+    bool M(object obj)
+    {
+        if ([||]obj is TestFile)
+        {
+            var file = (TestFile)obj;
+        }
+    }
+}");
+        }
+
+        [Fact, Trait(Traits.Feature, Traits.Features.CodeActionsInlineTypeCheck)]
+        public async Task TestNotOnNullable()
+        {
+            await TestMissingAsync(
+@"struct TestFile
+{
+    bool M(object obj)
+    {
+        if ([||]obj is TestFile?)
```

Comment:
❔ Is it true that it would be valid to replace this with `obj is TestFile file` *provided* `file` is never used directly (i.e. passed as an argument, assigned to an anonymous type variable)?

* Uses of `((TestFile?)obj).HasValue` become `true`
* Uses of `((TestFile?)obj).Value` become `file`
* Other uses of `((TestFile?)obj)` become `file`
* Uses of `((TestFile)obj)` become `file`